src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -248,26 +268,73 @@ def _maybe_warn_unowned(
+            "\n  * Multiple resolves. A target may exist for the import, but not share the same "
+            f"resolve of `{resolve}` from the target {address}. Pants only can infer dependencies "
+            f"on targets using the same resolve. You may need to set the `resolve=` field for "
+            f"whichever target provides that import (and possibly use the `parametrize` mechanism "
+            f"if that code should work with multiple resolves). See "
+            f"{doc_url('python-third-party-dependencies#multiple-lockfiles')}."
         )
-
-        if raise_error:
-            raise UnownedDependencyError(
-                "One or more unowned dependencies detected. Check logs for more details."
-            )
+        if project_uses_multiple_resolves
+        else ""
+    )
+    msg = softwrap(
+        f"""
+        Pants cannot infer owners for the following imports in the file {file} (from the target
+        {address}): 
+
+        {bullet_list(unowned_imports_with_lines)}
+
+        If you do not expect an import to be inferrable, add `# pants: no-infer-dep` to the 
+        import line.
+
+        Is the import from a third-party dependency (see
+        {doc_url("python-third-party-dependencies")})? Some common issues:
+
+          * Pants does not know about the requirement, i.e. there is not a `python_requirement`\
+          target for it. You can run `./pants filter --target-type=python_requirement ::` to see\
+          all `python_requirement` targets. If you are using `requirements.txt`, Poetry,\
+          or Pipenv, make sure you have `python_requirements`, `poetry_requirements`, and\
+          `pipenv_requirements` target generators, respectively. Double check that the requirement\
+          is in the respective file like `requirements.txt`.
+          * The third-party requirement exposes modules different than Pants's default, e.g.\
+          `ansicolors` exposing `colors`. Set the `modules` or `module_mapping` fields.
+          * Ambiguity. See below.{resolves_short_line}
+
+        Is the import from first-party code? Some common issues:
+
+          * The file does not exist.
+          * The file is missing an owning target like `python_sources`. Run `./pants list\
+          path/to/file.py` to see if there are any owners. If not, run `./pants tailor` to generate\
+          targets.
+          * The source root is not correctly set up, which is how Pants knows for example that the\
+          file `src/py/my_project/app.py` exposes the module `my_project.app`. See\
+          {doc_url('source-roots')}.
+          * Ambiguity. See below.{resolves_short_line}
+
+        Some common issues with both first and third-party imports:
+
+          * Ambiguity. See above logs if there are warnings and instructions.{resolves_long_line}
+        """
+    )
+    if unowned_dependency_behavior is UnownedDependencyUsage.LogWarning:
+        logger.warning(msg)
+    else:
+        raise UnownedDependencyError(msg)
```

Comment:
We before were logging at ERROR level, and then raising an error. I didn't see much benefit to splitting it up like that.